src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorCodeDocument.cs

```diff
@@ -182,8 +183,14 @@ internal RazorHtmlDocument GetHtmlDocument()
             return result;
         }
 
-        result = RazorHtmlWriter.GetHtmlDocument(this);
-        _properties.HtmlDocument.SetValue(result);
+        lock (_htmlDocumentLock)
+        {
+            if (!_properties.HtmlDocument.TryGetValue(out result))
```

Comment:
>if (!_properties.HtmlDocument.TryGetValue(out result)) [](http://example.com/codeflow?start=12&length=54)

This could just use a lazy instead of the lock, but that would require explicitly caching the htmldocument in this object. I assume that's ok, and I'll change this to use a lazy if code reviews indicate it's more desirable than an explicit lock.

Reply:
Personally, I'd have preferred a lock rather than store the `RazorHtmlDocument` in two separate places. IMO, this obscures the reason why `RazorHtmlDocument` is stored in the `PropertyTable` at all and could easily be removed accidentally, which might negatively impact the legacy editor.

Reply:
I'm not really opinionated, went ahead and switched back to the code from the first commit which used an explicit lock